### PR TITLE
F #-: Add support for Multi-Node deployment

### PR DESCRIPTION
**Pull request by Aletibazo**

### CODE REVIEW

File: appliances/MinIO/appliance.sh [hunk 1]
```diff
@@ -191,9 +194,34 @@ service_configure()
         chown -R minio-user:minio-user /opt/minio
     fi
 
-    # Edit MinIO environment variable file
-    if (( local_drives_count > 1 )); then
+
+    # If Multi-Node deployment, edit MinIO volumes
+    if [[ ${ONEAPP_MINIO_MULTI} =~ ^(yes|YES)$ ]]; then
+        local_subdomain=$(awk -F. '{ print $1 }' <<<${ONEAPP_MINIO_HOSTNAME})
```
Comment (sk4zuzu) [inline appliances/MinIO/appliance.sh:200]:
```suggestion
        local_subdomain=$(awk -F. '{ print $1 }' <<< "$ONEAPP_MINIO_HOSTNAME")
```

File: appliances/MinIO/appliance.sh [hunk 2]
```diff
@@ -191,9 +194,34 @@ service_configure()
...
+        local_domain=$(sed -e "s/[^.]*//" <<<"$ONEAPP_MINIO_HOSTNAME")
```
Comment (sk4zuzu) [inline appliances/MinIO/appliance.sh:201]:
```suggestion
        local_domain=$(sed -e "s/[^.]*//" <<< "$ONEAPP_MINIO_HOSTNAME")
```

File: appliances/MinIO/appliance.sh [hunk 3]
```diff
@@ -191,9 +194,34 @@ service_configure()
...
+        # Dynamically create hosts entries on /etc/hosts, getting IP for each node from onegate
+        # Do this before updating env file, so MinIO can connect to other nodes
+        # Get array of nodes IDs from MinIO role on service
+        local_nodes=($(onegate service show --json | jq -r '.SERVICE.roles[] | select(.name=="MinIO").nodes[].deploy_id'))
```
Comment (sk4zuzu) [inline appliances/MinIO/appliance.sh:205]: How about using something like this stuff below instead? :thinking: 

```
onegate service show -j --extended | jq -r '.SERVICE.roles[]|select(.name=="MinIO").nodes[].vm_info.VM.TEMPLATE.CONTEXT.ETH0_IP'
```

or

```
onegate service show -j --extended | jq -r '.SERVICE.roles[]|select(.name=="MinIO").nodes[].vm_info.VM.TEMPLATE.NIC.IP'
```

You get the same IPs with a single API call, the if you want to do a loop:

```
onegate service show -j --extended | jq -r '.SERVICE.roles[]|select(.name=="MinIO").nodes[].vm_info.VM.TEMPLATE.NIC.IP' | while read IP
; do echo "IP -> $IP"; done
```

**Review by sk4zuzu — CHANGES_REQUESTED** [review]

File: appliances/MinIO/appliance.sh [hunk 4]
```diff
@@ -349,6 +387,16 @@ MINIO_OPTS="--console-address :9001"
 EOF
 }
 
+edit_hosts()
+{
+    # add MinIO section to hosts
+    msg info "Add MinIO section to fstab"
```
Comment (rsmontero) [inline appliances/MinIO/appliance.sh:393]: to fstab?

File: appliances/MinIO/appliance.sh [hunk 5]
```diff
@@ -167,33 +174,63 @@ service_configure()
...
-        if [[ -z "${ONEAPP_MINIO_TLS_CERT}" ]] || [[ -z "${ONEAPP_MINIO_TLS_KEY}" ]]; then
-            msg info "Autogenerating TLS certificates..."
-            generate_tls_certs
+        if [[ -f "${local_minio_certs_path}/public.crt" ]] || [[ -f "${local_minio_certs_path}/private.key" ]]; then
+            msg info "Certificates already exist. Skipping."
         else
-            msg info "Configuring provided TLS certificates..."
-            echo ${ONEAPP_MINIO_TLS_CERT} | base64 --decode >> /opt/minio/certs/public.crt
-            echo ${ONEAPP_MINIO_TLS_KEY} | base64 --decode >> /opt/minio/certs/private.key
+            if [[ ! -d "${local_minio_certs_path}" ]]; then
```
Comment (rsmontero) [inline appliances/MinIO/appliance.sh:186]: is this needed? we can just mkdir -p without checking

File: appliances/MinIO/appliance.sh [hunk 6]
```diff
@@ -167,33 +174,63 @@ service_configure()
...
+
+            if [[ -z "${ONEAPP_MINIO_TLS_CERT}" ]] || [[ -z "${ONEAPP_MINIO_TLS_KEY}" ]]; then
+                msg info "Autogenerating TLS certificates..."
+                generate_tls_certs
+            else
+                msg info "Configuring provided TLS certificates..."
+                echo ${ONEAPP_MINIO_TLS_CERT} | base64 --decode >> /opt/minio/certs/public.crt
+                echo ${ONEAPP_MINIO_TLS_KEY} | base64 --decode >> /opt/minio/certs/private.key
+            fi
+            msg info "Give ownership of /opt/minio to minio-user"
```
Comment (rsmontero) [inline appliances/MinIO/appliance.sh:201]: Change ownership of /... to ....

File: appliances/MinIO/appliance.sh [hunk 7]
```diff
@@ -114,7 +121,7 @@ service_install()
 # Start MinIO-volumes
 # End MinIO-volumes
 EOF
-
+    edit_hosts
```
Comment (rsmontero) [inline appliances/MinIO/appliance.sh:124]: maybe we do not need a function? or merge the logic of 
```

       # For each node, get IP and create entry using MinIO Multi-Node notation
        i=1
        for node in "${local_nodes[@]}"
        do
            local_ip=$(onegate vm show ${node} --json | jq -r '.VM.TEMPLATE.NIC[].IP')
            local_hosts+="${local_ip} ${local_subdomain}$((i))${local_domain}\n"
            i=$((i + 1))
        done
        msg info "Adding MinIO hostnames to /etc/hosts"
        sed -i "/# End MinIO hosts/i ${local_hosts}" /etc/hosts
```

into the function like setup_etc_hosts

File: appliances/MinIO/appliance.sh (hunk 4 above)
Comment (OpenNebulaSupport) [reply appliances/MinIO/appliance.sh:393]: Copy-paste error :)

Comment (Aletibazo) [reply appliances/MinIO/appliance.sh:393]: Copy-paste error :)

**Review by rsmontero — APPROVED** [review]

**Pull request merged**